google/cloud/storage/internal/async/writer_connection_buffered.cc

```diff
@@ -0,0 +1,410 @@
+      " uploads using the same upload id. Most likely an application bug.",
+      std::move(eib)
+          .WithMetadata("gcloud-cpp.storage.upload_id", upload_id)
+          .WithMetadata("gcloud-cpp.storage.resend_offset", previous)
+          .WithMetadata("gcloud-cpp.storage.persisted_size", returned));
+}
+
+/**
+ * Implements an `AsyncWriterConnection` that automatically resumes and resends
+ * data.
+ *
+ * This class is used in the implementation of
+ * `AsyncClient::StreamingWriteObject()`. Please see that function for the
+ * motivation.
+ *
+ * This implementation of `AsyncWriterConnection` keeps an in-memory
+ * `resend_buffer_` of type `absl::Cord`. New data is added to the end of the
+ * Cord. Flushed data is removed from the front of the Cord.
+ *
+ * Applications threads add data by calling `Write()` and `Finalize()`.
+ *
+ * The buffer is drained by an asynchronous loop running in background threads.
+ * This loop starts (if needed) when new data is appended to the
+ * `resend_buffer_`. If the buffer is neither full nor approaching fullness
+ * the loop calls `impl_->Write()` to upload data to the service.
+ *
+ * When the application finalizes an upload the loop calls `impl_->Finalize()`
+ * and sends any previously buffered data as well as the new data.
+ *
+ * If the buffer is getting full, the loop uses `impl_->Flush()` instead of
+ * `impl_->Write()` to upload data, and it also queries the status of the upload
+ * after each `impl_->Flush()` call.
+ *
+ * If any of these operations fail the loop resumes the upload using a factory
+ * function to create new `AsyncWriterConnection` instances. This class assumes
+ * that the factory function implements the retry loop.
+ *
+ * If the factory function returns an error the loop ends.
+ *
+ * The loop also ends if there are no more bytes to send in the resend buffer.
+ */
+class AsyncWriterConnectionBuffered
+    : public storage_experimental::AsyncWriterConnection {
+ public:
+  explicit AsyncWriterConnectionBuffered(
+      WriterConnectionFactory factory,
+      std::unique_ptr<storage_experimental::AsyncWriterConnection> impl,
+      std::size_t buffer_size_lwm, std::size_t buffer_size_hwm)
+      : factory_(std::move(factory)),
+        buffer_size_lwm_(buffer_size_lwm),
+        buffer_size_hwm_(buffer_size_hwm),
+        impl_(std::move(impl)) {
+    auto state = impl_->PersistedState();
+    if (absl::holds_alternative<storage::ObjectMetadata>(state)) {
+      finalized_.set_value(
+          absl::get<storage::ObjectMetadata>(std::move(state)));
+      cancelled_ = true;
+      resume_status_ = internal::CancelledError("upload already finalized",
+                                                GCP_ERROR_INFO());
+      return;
+    }
+    buffer_offset_ = absl::get<std::int64_t>(state);
+  }
+
+  void Cancel() override {
+    std::unique_lock<std::mutex> lk(mu_);
+    cancelled_ = true;
+    lk.unlock();
+    return impl_->Cancel();
```

Comment:
Below you imply that accessing `impl_` needs synchronization.

Reply:
Fixed here and below.